File: github_wiki.py
import os
import subprocess
import shutil
from datetime import datetime
from dotenv import load_dotenv

load_dotenv()

# 1. 환경 변수 읽기
TOKEN = os.getenv("GITHUB_PAT")
USER = os.getenv("GITHUB_USERNAME") 
REPO = os.getenv("GITHUB_REPO_NAME") 

# 2. URL 조립
TEAM_WIKI_URL = f"https://{TOKEN}@github.com/{USER}/{REPO}.wiki.git"
WIKI_TEMP_DIR = "team_wiki_tmp"

def update_sidebar(wiki_dir, page_name_no_ext):
    """
    _Sidebar.md의 <details> 내부에 새 회의록 링크를 촘촘하게 추가합니다.
    """
    sidebar_path = os.path.join(wiki_dir, "_Sidebar.md")
    new_link = f"- [[{page_name_no_ext}]]"
    
    lines = []
    if os.path.exists(sidebar_path):
        with open(sidebar_path, "r", encoding="utf-8") as f:
            lines = f.readlines()
    
    # 중복 링크 체크
    if any(new_link in line for line in lines):
        print(f"ℹ️ 사이드바에 이미 '{page_name_no_ext}' 링크가 존재합니다.")
        return

    # 삽입 위치 찾기 (<summary>📅 회의록</summary> 바로 다음 줄)
    insert_index = -1
    for i, line in enumerate(lines):
        if "<summary>📅 회의록</summary>" in line:
            insert_index = i + 1
            break
    
    if insert_index != -1:
        # 상단 공백은 마크다운 문법 유지를 위해 필요하지만, 아래는 바로 붙여서 삽입
        # \n- [[링크]] 형식을 사용하여 기존 리스트와 밀착시킵니다.
        lines.insert(insert_index, "\n" + new_link)
    else:
        # 파일이 없거나 태그가 없는 경우 초기 구조 생성
        lines = [
            "### 🛠️ 기술 스택\n",
            "- [[Frontend]]\n",
            "- [[Backend]]\n\n",
            "<details>\n",
            "<summary>📅 회의록</summary>\n",
            f"\n{new_link}\n", # 첫 생성 시에도 문법 보호를 위해 줄바꿈 포함
            "\n</details>\n"
        ]

    with open(sidebar_path, "w", encoding="utf-8") as f:
        f.writelines(lines)
    print("📝 사이드바(_Sidebar.md) 리스트 밀착 업데이트 완료")

def update_team_wiki(summary_md_path):
    if not os.path.exists(summary_md_path):
        print(f"❌ 요약본 파일을 찾을 수 없습니다: {summary_md_path}")
        return

    if not os.path.exists(WIKI_TEMP_DIR):
        print("📂 팀 위키 클론 중...")
        subprocess.run(["git", "clone", TEAM_WIKI_URL, WIKI_TEMP_DIR], check=True)
    else:
        print("🔄 팀 위키 최신화 중...")
        subprocess.run(["git", "-C", WIKI_TEMP_DIR, "pull"], check=True)

    # 2. 파일 이름 결정 (날짜-회의록 형식 권장)
    today = datetime.now().strftime("%Y-%m-%d")
    # 실제 파일명을 '2026-03-16-회의록' 형태로 자동 생성
    wiki_page_name = f"{today}-회의록" 
    wiki_file_name = f"{wiki_page_name}.md"
    dest_path = os.path.join(WIKI_TEMP_DIR, wiki_file_name)
    
    shutil.copy(summary_md_path, dest_path)

    # 3. 사이드바 업데이트
    update_sidebar(WIKI_TEMP_DIR, wiki_page_name)

    # 4. Git Push
    print("🚀 팀 위키로 푸시 중...")
    try:
        subprocess.run(["git", "-C", WIKI_TEMP_DIR, "add", "."], check=True)
        status = subprocess.run(["git", "-C", WIKI_TEMP_DIR, "status", "--porcelain"], capture_output=True, text=True).stdout
        if not status:
            print("✨ 변경 사항이 없어 푸시를 건너뜁니다.")
            return

        subprocess.run(["git", "-C", WIKI_TEMP_DIR, "commit", "-m", f"docs: {today} 회의록 자동 업데이트"], check=True)
        subprocess.run(["git", "-C", WIKI_TEMP_DIR, "push"], check=True)
        print(f"✅ Wiki 업데이트 완료: {wiki_file_name}")
    except subprocess.CalledProcessError as e:
        print(f"⚠️ 에러 발생: {e}")

if __name__ == "__main__":
    # summarize.py가 생성한 최종 결과물 경로를 넣어주세요
    update_team_wiki("transcripts/meeting_summary.md")

File: github_wiki.js
require("dotenv").config();
const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");

// 환경 변수 읽기
const TOKEN = process.env.GITHUB_PAT;
const USER = process.env.GITHUB_USERNAME;
const REPO = process.env.GITHUB_REPO_NAME;

// [보안] 테스트 환경 모드
const TEST_MODE = process.env.TEST_MODE !== "false";

const WIKI_TEMP_DIR = "team_wiki_tmp";

function updateSidebar(wikiDir, pageNameNoExt) {
  const sidebarPath = path.join(wikiDir, "_Sidebar.md");
  const newLink = `- [[${pageNameNoExt}]]`;

  let content = "";
  if (fs.existsSync(sidebarPath)) {
    content = fs.readFileSync(sidebarPath, "utf-8");
  }

  // 중복 링크 체크
  if (content.includes(newLink)) {
    console.log(`ℹ️ 사이드바에 이미 '${pageNameNoExt}' 링크가 존재합니다.`);
    return;
  }

  // 1. 모든 줄바꿈을 단일화하고 앞뒤 공백 제거 (청소 작업)
  // filter(Boolean)으로 빈 줄을 일단 다 제거해서 '밀착' 상태로 만듭니다.
  let lines = content
    .split("\n")
    .map((l) => l.trim())
    .filter((line) => line !== "");

  // 2. 삽입 위치 찾기
  const summaryTag = "<summary>📅 회의록</summary>";
  const insertIndex = lines.findIndex((line) => line.includes(summaryTag));

  if (insertIndex !== -1) {
    /**
     * 핵심 로직:
     * - summaryTag 다음에 '빈 줄("")' 하나 넣고
     * - 그 다음에 '새 링크(newLink)'를 넣음
     * - 그 뒤에는 필터링된 기존 리스트들이 자동으로 붙음
     */
    lines.splice(insertIndex + 1, 0, "", newLink);
  } else {
    // 구조가 아예 없을 때 초기 생성
    lines = [
      "### 🛠️ 기술 스택",
      "- [[Frontend]]",
      "- [[Backend]]",
      "",
      "<details>",
      "<summary>📅 회의록</summary>",
      "",
      newLink,
      "</details>",
    ];
  }

  // 3. 최종 저장
  fs.writeFileSync(sidebarPath, lines.join("\n"), "utf-8");
  console.log("📝 사이드바(_Sidebar.md) 리스트 밀착 업데이트 완료");
}

function updateTeamWiki(summaryMdPath) {
  if (!fs.existsSync(summaryMdPath)) {
    console.log(`❌ 요약본 파일을 찾을 수 없습니다: ${summaryMdPath}`);
    process.exitCode = 1;
    return;
  }

  if (!TOKEN || !USER || !REPO) {
    console.log(
      "⏭️  GITHUB_PAT / GITHUB_USERNAME / GITHUB_REPO_NAME 중 하나 이상 미설정 → GitHub Wiki 업로드를 스킵합니다.",
    );
    return;
  }

  if (TEST_MODE) {
    console.log(
      `🚨 [TEST_MODE] 테스트 모드가 활성화되어 있습니다. 실제 Github Wiki 저장소에서 clone / push하지 않습니다.`,
    );
    return;
  }

  const teamWikiUrl = `https://${TOKEN}@github.com/${USER}/${REPO}.wiki.git`;

  if (!fs.existsSync(WIKI_TEMP_DIR)) {
    console.log("📂 팀 위키 클론 중...");
    execSync(`git clone ${teamWikiUrl} ${WIKI_TEMP_DIR}`, {
      stdio: "inherit",
    });
  } else {
    console.log("🔄 팀 위키 최신화 중...");
    execSync(`git -C ${WIKI_TEMP_DIR} pull`, { stdio: "inherit" });
  }

  const today = new Date();
  const pad = (num) => String(num).padStart(2, "0");
  const todayStr = `${today.getFullYear()}-${pad(today.getMonth() + 1)}-${pad(today.getDate())}`;

  const wikiPageName = `${todayStr}-회의록`;
  const wikiFileName = `${wikiPageName}.md`;
  const destPath = path.join(WIKI_TEMP_DIR, wikiFileName);

  fs.copyFileSync(summaryMdPath, destPath);

  // 사이드바 업데이트
  updateSidebar(WIKI_TEMP_DIR, wikiPageName);

  // Git Push
  console.log("🚀 팀 위키로 푸시 중...");
  try {
    execSync(`git -C ${WIKI_TEMP_DIR} add .`, { stdio: "inherit" });

    // 변경 사항 체크
    const status = execSync(`git -C ${WIKI_TEMP_DIR} status --porcelain`, {
      encoding: "utf8",
    });
    if (!status) {
      console.log("✨ 변경 사항이 없어 푸시를 건너뜁니다.");
      return;
    }

    execSync(
      `git -C ${WIKI_TEMP_DIR} commit -m "docs: ${todayStr} 회의록 자동 업데이트"`,
      { stdio: "inherit" },
    );
    execSync(`git -C ${WIKI_TEMP_DIR} push`, { stdio: "inherit" });
    console.log(`✅ Wiki 업데이트 완료: ${wikiFileName}`);
  } catch (e) {
    console.error(`⚠️ 에러 발생:`, e.message);
    process.exitCode = 1;
  }
}

if (require.main === module) {
  updateTeamWiki(path.join(__dirname, "transcripts", "meeting_summary.md"));
}

module.exports = { updateTeamWiki };
